### PR TITLE
Added error code 10 NACKs

### PR DESCRIPTION
**Pull request by AndyFlintNHS**
### CHANGES
- I've introduced the code 10 NACKs for:
    - DownloadError on sending EHR core
    - PresignedUrlNotFound Error on sending EHR core
    - DownloadError on sending fragments while handling continue request
    - PresignedUrlNotFound Error on sending fragments while handling continue request
    
- I've also separated PresignedUrlNotFoundError. There's now:
    - PresignedUrlNotFoundError (Which will send an error code 10 NACK when caught)
    - PresignedUrlNotFoundWhileDeletingEhrError (which doesn't need to send a NACK as it's part of the deletion process)

- Testing for error logging, it now verifies that it logs the 'error' if the error has an external cause, and logs the correct details when logging a new 'NegativeAcknowledgementError', verifying that it logs the correct details about the internal failure code

### CODE REVIEW

File: src/constants/enums.js [hunk 1]
```diff
@@ -94,6 +100,16 @@ export const AcknowledgementErrorCode = {
     internalErrorCode: `${Gp2gpError.CODE_06.errorCode}-B`,
     internalErrorDescription: 'Partial ingestion, cannot send a full EHR out'
```
Comment (chrisbloe-nhse) [inline src/constants/enums.js:101]: Should these messages also be parameterised?

File: src/errors/__tests__/errors.test.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  const negativeAcknowledgementErrors = [
+    {
+      type: PatientRecordNotFoundError,
+      acknowledgementErrorCode: AcknowledgementErrorCode.ERROR_CODE_06_A,
+      errorMessage: errorMessages.PATIENT_RECORD_NOT_FOUND_ERROR
+    },
+    {
+      type: PatientRecordNotFoundError,
+      acknowledgementErrorCode: AcknowledgementErrorCode.ERROR_CODE_06_B,
+      errorMessage: errorMessages.PATIENT_RECORD_NOT_FOUND_ERROR
```
Comment (chrisbloe-nhse) [inline src/errors/__tests__/errors.test.js:19]: This is the same message as `06-A`?

File: src/errors/__tests__/errors.test.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      acknowledgementErrorCode: AcknowledgementErrorCode.ERROR_CODE_10_A,
+      errorMessage: errorMessages.DOWNLOAD_ERROR
+    },
+    {
+      type: PresignedUrlNotFoundError,
+      acknowledgementErrorCode: AcknowledgementErrorCode.ERROR_CODE_10_B,
+      errorMessage: errorMessages.PRESIGNED_URL_NOT_FOUND_ERROR
+    }
+  ];
+  it.each(negativeAcknowledgementErrors)(
```
Comment (chrisbloe-nhse) [inline src/errors/__tests__/errors.test.js:32]: Additional new line, please.

File: src/services/transfer/__tests__/transfer-out-fragments.test.js [hunk 4]
```diff
@@ -228,7 +229,8 @@ describe('transferOutFragments', () => {
 
   it('should update fragment status to DOWNLOAD_FAILED when unable to retrieve fragment from presigned url', async () => {
     // given
-    const error = new DownloadError('error');
+    const acknowledgementErrorCode = AcknowledgementErrorCode.ERROR_CODE_10_A;
+    const error = new DownloadError(null, acknowledgementErrorCode);
```
Comment (chrisbloe-nhse) [inline src/services/transfer/__tests__/transfer-out-fragments.test.js:233]: Does a `null` parameter not cause a null pointer?

File: src/services/ehr-repo/__tests__/get-fragment.test.js [hunk 5]
```diff
@@ -60,23 +61,22 @@ describe('getFragment', () => {
...
+    expect(logError).toBeCalledWith(axios404Error)
   });
 
   it('should throw a DownloadError if it failed to download the EHR from the presigned URL', async () => {
     // given
     const inboundConversationId = '0E745F0A-1E8D-4C5F-AE13-B5758F331B29';
     const messageId = 'EBFAE96A-ED38-40C4-A854-58175D15EAEC';
-    const error = new DownloadError();
+    const acknowledgementErrorCode = AcknowledgementErrorCode.ERROR_CODE_10_A;
+    const error = new DownloadError(null, acknowledgementErrorCode);
```
Comment (chrisbloe-nhse) [inline src/services/ehr-repo/__tests__/get-fragment.test.js:79]: Does a `null` parameter not cause a null pointer?

File: src/services/handler/__test__/continue-message-handler.test.js [hunk 6]
```diff
@@ -217,4 +234,59 @@ describe('continueMessageHandler', () => {
...
+      failureReason: FailureReason.EHR_DOWNLOAD_FAILED
+    },
+    {
+      errorType: PresignedUrlNotFoundError,
+      acknowledgementErrorCode: AcknowledgementErrorCode.ERROR_CODE_10_B,
+      conversationStatus: conversationStatus.OUTBOUND_FRAGMENTS_SENDING_FAILED,
+      failureReason: FailureReason.MISSING_FROM_REPO
+    }
+  ];
+  it.each(testCasesForFragmentTransferErrors)(
```
Comment (chrisbloe-nhse) [inline src/services/handler/__test__/continue-message-handler.test.js:252]: Additional new line, please.

File: src/services/transfer/__tests__/transfer-out-ehr-core.test.js [hunk 7]
```diff
@@ -101,73 +115,74 @@ describe('transferOutEhrCore', () => {
...
+        failureReason: FailureReason.EHR_DOWNLOAD_FAILED
+      },
+      {
+        errorType: PresignedUrlNotFoundError,
+        acknowledgementErrorCode: AcknowledgementErrorCode.ERROR_CODE_10_B,
+        conversationStatus: conversationStatus.OUTBOUND_FAILED,
+        failureReason: FailureReason.MISSING_FROM_REPO
+      }
+    ];
+    it.each(testCasesForCoreTransferErrors)(
```
Comment (chrisbloe-nhse) [inline src/services/transfer/__tests__/transfer-out-ehr-core.test.js:136]: Additional new line, please!

File: src/services/transfer/__tests__/transfer-out-ehr-core.test.js [hunk 8]
```diff
@@ -101,73 +115,74 @@ describe('transferOutEhrCore', () => {
...
-      expect(logInfo).toHaveBeenCalledWith(logInfoMessages[0]);
-      expect(logInfo).toHaveBeenCalledWith(logInfoMessages[1]);
-      expect(updateConversationStatus).toHaveBeenCalledWith(
-        conversationId,
-        ConversationStatus.OUTBOUND_FAILED,
-        FailureReason.EHR_DOWNLOAD_FAILED
-      );
-      expect(logError).toHaveBeenCalledWith(errorMessages.DOWNLOAD_ERROR, undefined);
-    });
+    const testCasesForCoreTransferErrors = [
```
Comment (chrisbloe-nhse) [inline src/services/transfer/__tests__/transfer-out-ehr-core.test.js:118]: Missing `ERROR_CODE_06_B`?

**Review by chrisbloe-nhse — CHANGES_REQUESTED** [review]
Have left some comments, but generally looking good :).

File: src/constants/enums.js (hunk 1 above)
Comment (AndyFlintNHS) [reply src/constants/enums.js:101]: To me, this isn't something that'd ever be used outside of the context of this object, so I didn't really want to break it out any further.

File: src/services/ehr-repo/__tests__/get-fragment.test.js (hunk 5 above)
Comment (AndyFlintNHS) [reply src/services/ehr-repo/__tests__/get-fragment.test.js:79]: It doesn't. The only place that the 'error' parameter is used in this constructor has an if check around it before it's used

File: src/errors/__tests__/errors.test.js (hunk 2 above)
Comment (AndyFlintNHS) [reply src/errors/__tests__/errors.test.js:19]: Yes, the 'errorMessage' on the PatientRecordNotFoundError is the same for both the 06-A and 06-B. 

It's something I was wanting to take another look at at some point... have our internalErrorDescriptions replaced these?

File: src/services/transfer/__tests__/transfer-out-ehr-core.test.js (hunk 8 above)
Comment (AndyFlintNHS) [reply src/services/transfer/__tests__/transfer-out-ehr-core.test.js:118]: I can add it, but I'll explain why I missed this off:

The only place that the acknowledgementErrorCode is used in the test is to check that we're passing the gp2gpError to 'sendAcknowledgement'. 06-A and 06-B have the same gp2gpError and there's no need to update the database status so they'd just be an identical test.

File: src/services/transfer/__tests__/transfer-out-fragments.test.js (hunk 4 above)
Comment (AndyFlintNHS) [reply src/services/transfer/__tests__/transfer-out-fragments.test.js:233]: Same as above, this null parameter is only used in the constructor along with a null check, so it never causes problems.

**Review by chrisbloe-nhse — APPROVED** [review]
Yup, happy with this now :).

**Pull request merged**